libase/tds/channel.go

```diff
@@ -0,0 +1,418 @@
+package tds
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
```

Comment:
Could we use the standard error wrapping instead?

Reply:
I thought about that, but that wouldn't represent the 'reality' of errors, as the `multierror.Error` is used to collect errors when closing the `Channel`s and `Conn`, where `Channel`s are - when viewed from the error-wrapping hierarchy - should be on the same level. In turn a `Channel`s returned `multierror.Error` also contains any leftover queued packages and errors:

```
tds.Conn.Close error:
  -> tds.Channel1.Close error:
    -> teardown error
    -> leftover pkg1
    -> leftover pkg2
         ...
    -> leftover err1
  ->tds.Channel2.Close error:
    -> teardown error
      ...
```

If we were to wrap them directly we'd receive a long list of errors to produce more sophisticated error messages further up, e.g. to display `closed connection with x channels with y unprocessed packages and z unhandled errors`.

If that is actually useful or can be used is something we'll have to see when implemented the `database/sql` shim.

I'd also started to write a `libase.MultiError` but though that someone probably already wrote something like this and used hashicorps version for the time being - that module is also MPL licensed, so it is compatible with our license.